### DISCUSSION TITLE
Support cross-build toolchain integration

### DISCUSSION THREAD

**Discussion by zhongwm** [microsoft/vcpkg#18346] [Ideas]
My question is, how to integrate vcpkg in linux with cross build toolchain, as well as sysroot? 

example:

- build machine is fedora30
- cross build toolchain is x86_64-centos7-gnu

from https://github.com/microsoft/vcpkg quickstart says:

> git clone https://github.com/Microsoft/vcpkg.git
> cd vcpkg

PS> .\bootstrap-vcpkg.bat
Linux:~/$ ./bootstrap-vcpkg.sh

**Comment by heydojo** [comment]
> Support cross-build toolchain integration

Docker has made this pretty much a waste of development time (unless you can't do it with Docker. Which is very rare.)
You spin up a docker image to do your build and if you want to compile for x86_64-centos7-gnu, you spin up a [centos7 Docker](https://hub.docker.com/_/centos/) image (x64 can be used to compile x86) to run your builds on via your machine which might have docker and fedora 30 installed locally.
Online CI/CD services can also take the pain out of configuring Docker for you. So the barrier to entry for using a build service which uses Docker is low and the costs for Enterprise are very competitive.

The justification for cross compiling (unless it's for an obscure architecture) isn't really there any more. Neither is there much of a demand for x86 binaries either. YMMV. Try [Docker](https://www.docker.com/). I can recommend [Gitlab's Docker integration](https://docs.gitlab.com/ce/ci/quick_start/) as an online example. If you have a project here on Github, you could mirror it on Gitlab and use the CI/CD there to cross compile using a Docker image.

EDIT: They were even talking about this exact same topic on [HN](https://news.ycombinator.com/item?id=21553923) today.

**Comment by JackBoosY** [comment]
I  think we already have some cross compile triplet and some contributors are improving them.

**Comment by JackBoosY** [comment]
Example: https://github.com/microsoft/vcpkg/pull/17505.

**Comment by longnguyen2004** [comment]
Hmm, there aren't really any good options at the moment. I guess the only way is to modify `scripts/toolchain/linux.cmake` and add `CMAKE_C_COMPILER` and `CMAKE_CXX_COMPILER`, or create your own toolchain file and specify it as `VCPKG_CHAINLOAD_TOOLCHAIN_FILE`

For maintainers:
@dg0yt and I have discussed the idea of an extra variable, something like `VCPKG_AUTOTOOLS_TRIPLET`, which would be the usual Autotools triplet of the platform you're targeting (ex. `x86_64-linux-gnu` or `x86_64-centos7-gnu`). This way, we can reuse `linux.cmake`, and users who want to cross-compile can simply duplicate the current `x64-linux.cmake` file and modify `VCPKG_AUTOTOOLS_TRIPLET`. Is this doable?
(cc @Neumann-A)

**Comment by JackBoosY** [comment]
In my opinion, we should do more support for cross-environment at the vcpkg level.
